src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+    pub(super) fn new() -> io::Result<Self> {
+        // Make sure AFD is able to be used.
+        if let Err(e) = afd::NtdllImports::force_load() {
+            return Err(crate::unsupported_error(format!(
+                "Failed to initialize unstable Windows functions: {}\nThis usually only happens for old Windows or Wine.",
+                e
+            )));
+        }
+
+        // Create a single AFD to test if we support it.
+        let afd = match Afd::new() {
+            Ok(afd) => afd,
+            Err(e) => return Err(crate::unsupported_error(format!(
+                "Failed to initialize \\Device\\Afd: {}\nThis usually only happens for old Windows or Wine.",
+                e,
+            ))),
+        };
+
+        let port = IoCompletionPort::new(0)?;
+        port.register(&afd, true)?;
+
+        log::trace!("new: handle={:?}", &port);
+
+        Ok(Poller {
+            port,
+            afd: Mutex::new(vec![Arc::new(afd)]),
+            sources: RwLock::new(HashMap::new()),
+            pending_updates: ConcurrentQueue::bounded(1024),
+            polling: AtomicBool::new(false),
+            packets: Mutex::new(Vec::with_capacity(1024)),
+            notifier: Arc::pin(
+                PacketInner::Wakeup {
+                    _pinned: PhantomPinned,
+                }
+                .into(),
+            ),
+        })
+    }
+
+    /// Whether this poller supports level-triggered events.
+    pub(super) fn supports_level(&self) -> bool {
+        true
+    }
+
+    /// Whether this poller supports edge-triggered events.
+    pub(super) fn supports_edge(&self) -> bool {
+        false
+    }
+
+    /// Add a new source to the poller.
+    pub(super) fn add(&self, socket: RawSocket, interest: Event, mode: PollMode) -> io::Result<()> {
+        log::trace!(
+            "add: handle={:?}, sock={}, ev={:?}",
+            self.port,
+            socket,
+            interest
+        );
+
+        // We don't support edge-triggered events.
+        if matches!(mode, PollMode::Edge) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "edge-triggered events are not supported",
+            ));
+        }
+
+        // Create a new packet.
+        let socket_state = {
+            let state = SocketState {
+                socket,
+                base_socket: base_socket(socket)?,
+                interest,
+                interest_error: true,
+                afd: self.afd_handle()?,
+                mode,
+                waiting_on_delete: false,
+                status: SocketStatus::Idle,
+            };
+
+            Arc::pin(IoStatusBlock::from(PacketInner::Socket {
+                packet: UnsafeCell::new(AfdPollInfo::default()),
+                socket: Mutex::new(state),
+            }))
+        };
+
+        // Keep track of the source in the poller.
+        {
+            let mut sources = self.sources.write().unwrap_or_else(|e| e.into_inner());
```

Comment:
instead of doing such a weird dance around poisoning all the time in this code, wouldn't it be more appropriate to either pull in a library that already implements locking in such a way (e.g. `parking-lot`) or write an abstraction around `RwLock` (and maybe `Mutex`) once, that does this, and use it in all this code?

Reply:
Hmm, yeah, it does look pretty ugly. I've added a `lock!` macro in order to do it automatically.